### ISSUE TITLE
Cannot make encrypted connection from rails app to SQL server 

### ISSUE THREAD

**Issue by amf9t2** [rails-sqlserver/tiny_tds#286]
Having problems? Have you checked these first:

* Are you using FreeTDS 0.95.80 or later? Check `$ tsql -C` to find out.
      Yes -  Version: freetds v1.00
* If not, please update then uninstall the TinyTDS gem and re-install it.
* Are you using Ubuntu? If so, you may have forgotten to install FreeTDS first.
      No. I am using Mac.

I am using tiny_tds(~> 0.7) gem successfully in rails application. I don't have any issues for doing sql operations. 

Now I want my session/connection to be encrypted. Below are the steps I tried to compile Freetds with openssl but no luck. I would appreciate if you could suggest or identify any thing wrong in this.

1. Executed rake command referring https://github.com/rails-sqlserver/tiny_tds#using-miniportile
 "rake TDSVER='7.1' TINYTDS_FREETDS_VERSION='0.95' -- --enable-system-openssl"

       Verifying by 
        - creating TDS client and executing query through rails console of same application.
              => client = TinyTds::Client.new username: 'username', password: 'password', host: 'host', encryption:    'require', database: 'dbname'  
              => r = client.execute("SELECT encrypt_option FROM sys.dm_exec_connections WHERE session_id = @@SPID")
               => r.each
                Output is [{"encrypt_option"=>"FALSE"}].

        - using wireshack and I can read the whole query in TDS Query Packet. 

2. Installed freetds manually by running "brew install freetds --with-openssl=DIR"
  Repeated same steps for verification but no luck.

Thanks for your time in advance.

**Comment by metaskills** [comment]
The miniportile install is mostly for compiling Windows static binaries. It can not, nor should not, replace you native platforms package manager. So number 2 is preferred. I would need more info to help. But I would have to Google search for most of my answers since I am not up on Homebrew.

**Comment by amf9t2** [comment]
Thanks for the instant response.

Any specific information you require ?

I am not getting any error doing this but its just that the queries and response are not encrypted inside packet even after installing freetds with openssl.

I tried another way to install freetds => "brew install freetds --HEAD --universal" and below is the output of tsql -C after installing. 
I also tried uninstalling and installing tiny_tds but no luck.
 
[TinyTds][v1.0.4][tsql]: /usr/local/bin/tsql
Compile-time settings (established with the "configure" script)
               Version: freetds v1.1.dev.20160617
             freetds.conf directory: /usr/local/Cellar/freetds/HEAD/etc
     MS db-lib source compatibility: no
        Sybase binary compatibility: no
                      Thread safety: yes
                      iconv library: yes
                        TDS version: 7.3
                              iODBC: no
                           unixodbc: no
              SSPI "trusted" logins: no
                           Kerberos: no
                            OpenSSL: yes
                             GnuTLS: no
                               MARS: n

**Comment by metaskills** [comment]
Hmmm... I had no idea Homebrew is using FreeTDS v1.1.development. We have not tested with that vision. Can you find a way to make Homebrew use the v1.0 release please?

**Comment by amf9t2** [comment]
Ok.

I uninstalled FreeTDS v1.1.development and installed FreeTDS v1.00 ("brew install freetds --universal --with-openssl=DIR"). Also uninstalled tiny_tds and installed but no luck. Am I missing any config file or do I have to add any extra changes so that tiny_tds refers that and enable encryption ? 

Below is the output of tsql -C.

[TinyTds][v1.0.4][tsql]: /usr/local/bin/tsql
Compile-time settings (established with the "configure" script)
Version: freetds v1.00
freetds.conf directory: /usr/local/Cellar/freetds/1.00/etc
MS db-lib source compatibility: no
Sybase binary compatibility: no
Thread safety: yes
iconv library: yes
TDS version: 7.3
iODBC: no
unixodbc: no
SSPI "trusted" logins: no
Kerberos: no
OpenSSL: yes
GnuTLS: no
MARS: no

Thanks.

**Comment by metaskills** [comment]
WOW, this looks good. Can we move to the showing me output from a simple IRB session and trying to make TinyTds::Client connect?

**Comment by amf9t2** [comment]
Sure. 
Below the is the new client and query output.

M00000:XXXX_service User$ rails c
Loading development environment (Rails 4.2.5)
[1] pry(main)> client = TinyTds::Client.new username: ‘username', password: ‘password', host: ‘host', encryption: 'require', database: ‘db name'

=> #<TinyTds::Client:0x007fbfc6d1cb20 @query_options={:as=>:hash, :symbolize_keys=>false, :cache_rows=>true, :timezone=>:local, :empty_sets=>true}>

[2] pry(main)> r = client.execute("SELECT encrypt_option FROM sys.dm_exec_connections WHERE session_id = @@SPID") 
=> #<TinyTds::Result:0x007fbfc6b8dc00 @query_options={:as=>:hash, :symbolize_keys=>false, :cache_rows=>true, :timezone=>:local, :empty_sets=>true}>

[3] pry(main)> r.each
=> [{"encrypt_option"=>"FALSE"}]

Thanks.

**Comment by metaskills** [comment]
There is no `encryption: 'require'` option for TinyTDS. We rely on the database telling the client to use an encrypted protocol, just like Azure does. Can you Google around for how to make SQL Server do that and report back?

**Comment by amf9t2** [comment]
I verified that our sql server doesn't have Force Encryption Enable but according to below link "When the ForceEncryption option for the Database Engine is set to No, encryption can be requested by the client application but is not required."
https://technet.microsoft.com/en-us/library/ms189067(v=sql.105).aspx

As per our use case we need SSL connection for our rails application but there are other applications where we connect non-SSL. 
Is there any way for rails application to make it configurable with the default to rely on database indicating it ?
Otherwise we'll need to fork or force SSL.

**Comment by metaskills** [comment]
> Is there any way for rails application to make it configurable with the default to rely on database indicating it ?

By that you are asking if TinyTDS can do it and hence is it in the FreeTDS/DBLIB spec. From the FreeTDS side, I believe others have done this by using the `freetds.conf` file. See [docs here](http://www.freetds.org/userguide/freetdsconf.htm). Perhaps this is where you got confused and tried to place the `encryption = require` in TinyTDS vs that file. Please try that first, I see in your output that the conf file is in looked up in  `/usr/local/Cellar/freetds/1.00/etc` but it can actually be in a few places if you Google more on FreeTDS's site. I believe this would work for you.

Ideally, we have always wanted a DBLIB login bit to set to tell SQL Server we want a required encrypted session. Previously that was not available in FreeTDS' DBLIB implementation which we use. Will take another look. But please do push ahead with the conf file.

**Comment by amf9t2** [comment]
Thanks for the answer. 

Added a link of "/usr/local/Cellar/freetds/1.00/etc/freetds.conf" file in home directory(~/.freetds.conf). Updated with below configuration. Still not sure if tiny_tds is referring .freetds.conf file. "encrypt_option"  output is still same.

[global]
	tds version = 7.0
	timeout = 10
	connect timeout = 30
	dump file = /tmp/freetds.log
	debug flags = 0xffff
	text size = 64512

# A typical Microsoft server
[ServerName]
	host = hostname
	port = default
	encryption = require

**Comment by metaskills** [comment]
I would add it to global unless you are using the `dataserver` "ServerName" in your TinyTDS connection.

**Comment by amf9t2** [comment]
Make sense.

Updated:
[global]
     tds version = 7.0
     timeout = 10
     connect timeout = 30
     dump file = /tmp/freetds.log
     debug flags = 0xffff
     host = hostname
     port = 1433
     encryption = require

Added all in global. Now looking at freetds log file and dump I can see tinytds uses freetds.conf file to get connection settings and when it tries to connect MSSQL server it fails. I assume there is something wrong on our server may be it is currently supporting "self signed cert " because commenting encryption flag in freetds.conf makes successful connection.

Next I am gonna look our server side for this. One more thing I would like to share is the output of 
tsql -S servername. Not sure what the problem is.

output:
M000000:xxxx_service user$ tsql -S servername
[TinyTds][v1.0.4][tsql]: /usr/local/bin/tsql
locale is "en_US.UTF-8"
locale charset is "UTF-8"
using default charset "UTF-8"
Error 20002 (severity 9):
	Adaptive Server connection failed
There was a problem connecting to the server

Thanks.

**Comment by amf9t2** [comment]
Wooow...We got cert fixed on our server.

 Also I am able to fix the Freetds connection and can verify the "encrypt_option" set to true for a session from "sys.dm_exec_connections" table.

This is the latest freetds.conf file. (we use MS SQL Server 2012)
[global]
	tds version = 7.4   
	timeout = 10
	connect timeout = 30
	dump file = /tmp/freetds.log
	debug flags = 0xffff
	host = hostname
	port = 1433
	encryption = require

Create connection command:
tsql -S servername:port -U 'servername\username' -P password

This is the output:
[TinyTds][v1.0.4][tsql]: /usr/local/bin/tsql
locale is "en_US.UTF-8"
locale charset is "UTF-8"
using default charset "UTF-8"
1> SELECT encrypt_option FROM sys.dm_exec_connections WHERE session_id = @@SPID;
2> go
encrypt_option
TRUE
(1 row affected)

Question: Still tinytds is not able to make encrypted connection. I doubt there is a difference in connection parameter between freetds and tinytds. Is it ?

**Comment by metaskills** [comment]
@amf9t2 Great stuff so far. Can you try `tds version = 7.3`? I did a few tests and found that things pass for me (unencrypted) using either 7.3 or 7.4. Just curious.

> I doubt there is a difference in connection parameter between freetds and tinytds. Is it ?

See this [old thread](http://lists.ibiblio.org/pipermail/freetds/2011q2/026878.html) ... so my understanding is no. That is backed up by the fact that it work with Azure, but that is forced encryption, so I am open to learning more. But from [what I can tell](https://github.com/FreeTDS/freetds/blame/master/include/sybdb.h#L1276-L1277) those settings in FreeTDS' DBLIB interface are no-ops and have been there since 2002 and means that the conf file is the only way and __SHOULD__ work.

Can you walk me thru how to turn SSL for my local SQL Server 2014 to see if I can help debug?

**Comment by amf9t2** [comment]
I tried changing tds version in tiny_tds but for some reason the tds version is not updated. If I don't set encryption property(freetds.conf) then tiny_tds make a successful connection with tds version 7.1 though I have set tds_version to 7.3. If I set encryption property then it breaks saying 'TinyTds::Error: Adaptive Server connection failed'.

I have noted one more thing is tiny_tds still referring FreeTDS 0.91.112 version I guess which is its default. Is it the expected behavior ?

From freetds.log:
 "11:02:17.402989 82608 (log.c:196):Starting log file for FreeTDS 0.91.112
	on 2016-06-23 11:02:17 with debug flags 0xffff.
11:02:17.403011 82608 (iconv.c:330):tds_iconv_open(0x7fd0ddeac910, UTF-8)
11:02:17.403018 82608 (iconv.c:349):setting up conversions for client charset "UTF-8 ......."

tdsconfig.log  :
Final connection parameters:
       server_name = servername:1433
       server_host_name = hostname
       ip_addr = XX.XXX.XXX.XXX
       instance_name = 
       port = 1433
       major_version = 7
       minor_version = 1
       block_size = 0
       language = us_english
       server_charset = 
       connect_timeout = 30
       client_host_name = M00000
       client_charset = UTF-8
       app_name = TinyTds
       user_name = server\username
       library = DB-Library
       bulk_copy = 0
       suppress_language = 0
       encrypt level = 2
       query_timeout = 10
       database = 
       dump_file = /tmp/freetds.log
       debug_flags = ffff
       text_size = 64512
       broken_dates = 0
       emul_little_endian = 0
       server_realm_name = 


I would be more than happy to walk through the steps but unfortunately that part was taken care by db team. I will get in touch with them for this but mean time I found this Microsoft link which can be of help (I guess you have already seen this ) https://support.microsoft.com/en-us/kb/316898 

Thanks.

**Comment by amf9t2** [comment]
Our db team recommended this steps : https://technet.microsoft.com/en-us/library/ms189067(v=sql.105).aspx

I hope it will be helpful.

**Comment by amf9t2** [comment]
Hey any luck with debugging this issue on your side ?

Thanks.

**Comment by amf9t2** [comment]
Hey,
I hope everything is good. 

I was thinking if this require additional development in tiny_tds gem then we would be happy to contribute. Please let me know what is your view on this.

Thanks.

**Comment by metaskills** [comment]
@amf9t2 Thanks so much for the offer. I am putting my energies to to the Rails v5 adapter work for the next month or so. So any help would be appreciated!

Given [my comment](https://github.com/rails-sqlserver/tiny_tds/issues/286#issuecomment-227924260) above? The author of FreeTDS states:

> To affect encryption in a db-lib program, your only
> option is to use the encryption property in freetds.conf.

To my knowledge no one has extended the DB-LIB interfaces a James Lowden suggested which means TinyTDS has no options other than to help document how to use the FreeTDS conf file to achieve the desired results. Thoughts?

**Comment by amf9t2** [comment]
No problem.

I will start understanding the functional workflow of gem and see how we can achieve the encrypted connection when MSSQL server is not set to forcefull encryption. The whole plan is to enable this and create a pull request. I would appreciate any input or guidance to direct me in correct files or def.

Thanks.

**Comment by metaskills** [comment]
@amf9t2 Thanks, did you read my comment above yours?

**Comment by amf9t2** [comment]
Ohhh.. I see ..
Question => 
"So I was testing the latest 0.91 with "Forced Encryption" turned on at the server level. And I monitored my network traffic and saw that indeed I was getting a secure connection. My question is, is there a way to specify this at the client level? In my code I have:

DBSETLDBNAME(login)

If the user specifies a :secure option. But it will not encrypt, only when I set the server to force clients. I can verify that my DBLIB connection is in plain text by doing a test query like "SELECT N'secret' AS [password]". Is this expected behavior? What should DBSETLDBNAME do?
 - Ken"

Answer => 
"DBSETLDBNAME(), as it should, simply sets the default database in the
login packet, obviating the need to call dbuse().  It has nothing to do
with encryption.  

To affect encryption in a db-lib program, your only option is to use
the encryption property in freetds.conf.  It can be set per-server or
globally for the client, but not programmatically per connection.  

To change that would mean extending db-lib, adding

	DBSETLENCRYPT( LOGINREC*, TDS_ENCRYPTION_LEVEL)

and/or adding an environment variable.  (DB-Library has no equivalent
of a connection string, but environment variables do roughly the same
job.) "

_I am still wondering that Freetds allows us to make encrypted connection (setting encryption in [global]) even when MSSQL server is not set to enforce encrypted connection. 
Is it because we have set encryption under global in conf file? 

If I understand right, to fix this issue we have to extend db-lib instead for tiny-tds ?

**Comment by metaskills** [comment]
> Is it because we have set encryption under global in conf file?

It could be. If you want to force it for a certain connection only, then you would need to create a named section (called dataserver) in the freetds.conf and tell TinyTDS to use that dataserver name.

> If I understand right, to fix this issue we have to extend db-lib instead for tiny-ads ?

That would be optional since FreeTDS is clearly saying that we can achieve the same results by using the freetds.conf file properly and TinyTDS will honer that since it is after all, just a Ruby wrapper around FreeTDS.

In summary, the words "fix this issues" is not correct. If you want to "extend and make FreeTDS more configurable" then yes, someone that feels that pain (maybe you) can start a conversation with the FreeTDS group and do one of the options James suggested. Then we can use said option. Hope that makes sense.

**Comment by amf9t2** [comment]
Greetings of the day,

The name section (called dataserver) approach worked. I had to set [server] specific configuration  and remove the encrypt option from [global].

File: .freetds.conf 

# A typical server
[<servername>]
  host = <hostname>
  port = 1433
  tds version = 7.3
  encryption = require

Rails application => Database.yml => add dataserver value.
dataserver: <server name>

One more thing I did but not sure if it helped me : 
I uninstalled tiny_tds and installed using below command.
"sudo gem install tiny_tds -- --with-freetds-include=/usr/local/include --with-freetds-lib=/usr/local/lib --with-iconv-include=/usr/local/Cellar/libiconv/1.14/include --with-iconv-lib=/usr/local/Cellar/libiconv/1.14/lib"

Yes you are right, I should not use the term "issue". My bad. 

I appreciate your time and patience.

Thanks. :) :)

**Comment by amf9t2** [comment]
Just an update for above comment. We don't want to run "sudo gem install tiny_tds ....." as "gem tiny_tds 1.0.4" inside gemfile will take care of this.

Thanks
